packages/app-mobile/components/NotesBar.tsx

```diff
@@ -0,0 +1,310 @@
+				paddingRight: 8,
+			},
+			searchIcon: {
+				fontSize: 22,
+			},
+			searchInput: {
+				alignItems: 'center',
+				backgroundColor: theme.backgroundColor,
+				paddingLeft: 8,
+				borderRadius: 4,
+				borderWidth: 1,
+				borderColor: theme.dividerColor,
+				height: 42,
+				flex: 1,
+			},
+			button: {
+				height: 42,
+				width: 42,
+				backgroundColor: theme.color4,
+				alignItems: 'center',
+				justifyContent: 'center',
+				borderRadius: 4,
+				flex: 0.5,
+				marginLeft: 8,
+			},
+			buttonIcon: {
+				color: theme.backgroundColor,
+				fontSize: 22,
+			},
+			inputGroup: {
+				justifyContent: 'space-between',
+			},
+		};
+
+		return StyleSheet.create(styles);
+	};
+
+	const titleComp = (
+		<View style={[styles().title, styles().horizontalFlex]}>
+			<Icon name='md-document'style={[styles().top, styles().titleIcon]} />
+			<Text style={[styles().top, styles().titleText]}>{_('Notes')}</Text>
+		</View>
+	);
+
+	const dividerComp = (
+		<View style={styles().divider}></View>
+	);
+
+	const handleNotesBarClose = () => {
+		props.dispatch({ type: 'NOTES_BAR_CLOSE' });
+	};
+
+	const closeButtonComp = (
+		<TouchableOpacity onPress={handleNotesBarClose}>
+			<Icon name="close" style={[styles().top, styles().closeIcon]}/>
+		</TouchableOpacity>
+	);
+
+	const renderIconButton = (icon: JSX.Element, onPress: ()=> Promise<void>) => {
+		return (
+			<TouchableOpacity style={styles().button} activeOpacity={0.8} onPress={onPress}>{icon}</TouchableOpacity>
+		);
+	};
+
+
+	const handleNewNote = async (isTodo: boolean) => {
+		let folderId = props.selectedFolderId !== Folder.conflictFolderId() ? props.selectedFolderId : null;
+		if (!folderId) folderId = props.activeFolderId;
+
+		props.dispatch({
+			type: 'NAV_BACK',
+		});
+
+		const newNote = await Note.save({
+			parent_id: folderId,
+			is_todo: isTodo ? 1 : 0,
+		}, { provisional: true });
+
+		props.dispatch({
+			type: 'NAV_GO',
+			routeName: 'Note',
+			noteId: newNote.id,
+		});
+	};
+
+	const addNoteButtonComp = renderIconButton(<Icon name='document-text-outline' style={styles().buttonIcon} />, () => handleNewNote(false));
+	const addTodoButtonComp = renderIconButton(<Icon name='checkbox-outline' style={styles().buttonIcon} />, () => handleNewNote(true));
+
+	const topComp = (
+		<View>
+			<View style={[styles().topContainer, styles().horizontalFlex]}>
+				{titleComp}
+				{closeButtonComp}
+			</View>
+			{dividerComp}
+		</View>
+	);
+
+	// Copied from './screens/search.js' and modified
+	const refreshSearch = async () => {
+		let notes_ = [];
+
+		if (query) {
+			if (props.settings['db.ftsEnabled']) {
+				notes_ = await SearchEngineUtils.notesForQuery(query, true);
+			} else {
+				const p = query.split(' ');
+				const temp = [];
+				for (let i = 0; i < p.length; i++) {
+					const t = p[i].trim();
+					if (!t) continue;
+					temp.push(t);
+				}
+
+				notes_ = await Note.previews(null, {
+					anywherePattern: `*${temp.join('*')}*`,
+				});
+			}
```

Comment:
This is almost the same as in `search.js`. Can this be factored out?

Reply:
`search.js` is a class component, so no.

Reply:
What? Of course you need to refactor this, we can't have duplicate code like this.

Reply:
I don't think it can be helped in this case. The code in `search.js`  uses `this.props` as opposed to `props` because it's a class component. There are also lines of code that are specific to `search.js` so it's not really reusable. To create a reusable function I would have to check if 'this.props' exists and then use 'props'  instead if it doesn't exist, same for 'this.setState', there's no way to tell how the setState function in the function component will be named, since it could be named anything. The function would be really messy. I think it's best left as is.

Reply:
>  The code in search.js uses this.props as opposed to props because it's a class component.

Do you seriously think there's no way to solve this? I'm not asking this to bother you - refactoring code is very important to avoid duplicate code all over the codebase. You're not the first one to do that, and if we were to allow this, the codebase would be unmaintainable - any small change we make somewhere will have to copied and pasted in many other places (which of course won't be done).

In this particular case, all you need to do is move the code from search.js to a separate function, replace this.props.xxx with function arguments, make the function return the notes. Call it from search.js and from your new code. setState and so on stay in the component.

Reply:
I actually thought there was no way to solve this. Thanks, I'll work on fixing it.